pkg/volumeclaimbinder/persistent_volume_claim_binder.go

```diff
@@ -0,0 +1,301 @@
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				volume := obj.(*api.PersistentVolume)
+				volumeIndex.Indexer.Add(volume)
+				syncVolume(binderClient, volume)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oldVolume := oldObj.(*api.PersistentVolume)
+				newVolume := newObj.(*api.PersistentVolume)
+				volumeIndex.Indexer.Update(newVolume)
+				if updateRequired(oldVolume, newVolume) {
+					syncVolume(binderClient, newVolume)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				volume := obj.(*api.PersistentVolume)
+				volumeIndex.Indexer.Delete(volume)
+			},
+		},
+	)
+	_, claimController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return kubeClient.PersistentVolumeClaims(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kubeClient.PersistentVolumeClaims(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&api.PersistentVolumeClaim{},
+		syncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				claim := obj.(*api.PersistentVolumeClaim)
+				syncClaim(volumeIndex, binderClient, claim)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				newClaim := newObj.(*api.PersistentVolumeClaim)
+				if newClaim.Status.VolumeRef == nil {
+					syncClaim(volumeIndex, binderClient, newClaim)
+				}
+			},
+		},
+	)
+
+	binder := &PersistentVolumeClaimBinder{
+		volumeController: volumeController,
+		claimController:  claimController,
+		volumeIndex:      volumeIndex,
+		client:           binderClient,
+	}
+
+	return binder
+}
+
+func updateRequired(oldVolume, newVolume *api.PersistentVolume) bool {
+	// Spec changes affect indexing and sorting volumes
+	if !reflect.DeepEqual(oldVolume.Spec, newVolume.Spec) {
+		return true
+	}
+	if !reflect.DeepEqual(oldVolume.Status, newVolume.Status) {
+		return true
+	}
+	return false
+}
+
+func syncVolume(binderClient binderClient, volume *api.PersistentVolume) (err error) {
+	glog.V(5).Infof("Synchronizing PersistentVolume[%s]\n", volume.Name)
+
+	if volume.Spec.ClaimRef != nil {
+		if volume.Status.Phase == api.VolumeAvailable {
+			volume.Status.Phase = api.VolumeBound
+			_, err := binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv.status: %v\n", err)
+			}
+		}
+
+		// verify the volume is still claimed by a user
+		if claim, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name); err == nil {
+			glog.V(5).Infof("PersistentVolume[%s] is bound to PersistentVolumeClaim[%s]\n", volume.Name, volume.Spec.ClaimRef.Name)
+			// rebuild the Claim's Status as needed
+			if claim.Status.VolumeRef == nil {
+				syncClaimStatus(binderClient, volume, claim)
+			}
+		} else {
+			//claim was deleted by user.
```

Comment:
I don't think it's safe to assume that without checking the error type (e.g. `errors.IsNotFound`).